### DISCUSSION TITLE
楽観排他について

### DISCUSSION THREAD

**Discussion by gikeihi** [domaframework/doma#841] [General]
現状、楽観排他はアノテーション`@Version`を利用して、ある数字型のコラムで実現しましたが、
ソースを見ると、毎回更新時、自動的に＋１しています。
timestamp型のコラムでも実現すればいいかなと思います。

例えば

@VersionTime
LocalDateTime update_datetime;

更新SQLは
```
update
  table
set
  update_datetime = CURRENT_TIMESTAMP()
where
  update_datetime = '2015-03-13 21:01:59'
```

**Comment by nakamura-to** [comment]
timestampベースの楽観的排他制御も技術的には可能ですが、数値ベースの方が効率的かつ確実なのでDomaでは数値ベースの楽観的排他制御のみを提供しています。
今のところtimestampベースの楽観的排他制御を実装する予定はありません。

**Reply by gikeihi** [reply]
早速のご返答ありがとうございます。
既存システムの改修ですが、ほとんどのテーブルには「update_datetime」のようなカラムがありますが、バージョンナンバーを使うと、コラムの追加が必要です。

実装の予定がないとは承知いたしましたが、何らかの方法で拡張することが可能ですが？

**Reply by nakamura-to** [reply]
`QueryImplementors`の実装を工夫することで拡張できると思われます。`QueryImplementors`ではUPDATE文やDELETE文の組み立てに介入できるので、「update_datetime」のようなカラムを意識したSQLを組み立てられるかもしれません。

独自に作成した`QueryImplementors`の実装は`Config`の実装クラスのメソッドで返すようにしてください。
https://github.com/domaframework/doma/blob/2.51.1/doma-core/src/main/java/org/seasar/doma/jdbc/Config.java#L106

**Reply by gikeihi** [reply]
ご教授、ありがとうございます。
試します。
